From 68cfd6923d2e62018b0556f5383d307b21a8038c Mon Sep 17 00:00:00 2001
From: Thibaud Michaud <thibaudm@chromium.org>
Date: Tue, 21 May 2024 16:22:01 +0200
Subject: [PATCH] Merged: [wasm][jspi] Fix promising export of a re-exported JS import

In this case the ref is not the instance but a WasmApiFunctionRef, so
the cast is invalid. Use the instance field instead.

R=jkummerow@chromium.org

Bug: 340102752

(cherry picked from commit 52d9e6db7f4664f0a0a11faaff4e7d28aa898bd8)

Change-Id: I8639cf154645c46ee0be66b31a5dd63c165e1f53
Reviewed-on: https://chromium-review.googlesource.com/c/v8/v8/+/5553793
Commit-Queue: Thibaud Michaud <thibaudm@chromium.org>
Reviewed-by: Jakob Kummerow <jkummerow@chromium.org>
Cr-Commit-Position: refs/branch-heads/12.5@{#22}
Cr-Branched-From: 15b9756484d5bda98ba273ae13f8db58200db4db-refs/heads/12.5.227@{#1}
Cr-Branched-From: 497d8573dc80b1b69052a834bec894cf5d4238e7-refs/heads/main@{#93350}
---

--- a/v8/src/wasm/wasm-js.cc
+++ b/v8/src/wasm/wasm-js.cc
@@ -2270,7 +2270,7 @@
     }
     i::Handle<i::WasmTrustedInstanceData> trusted_instance_data(
         i::WasmTrustedInstanceData::cast(
-            data->func_ref()->internal(i_isolate)->ref()),
+            data->instance()->trusted_data(i_isolate)),
         i_isolate);
     int func_index = data->function_index();
     i::Handle<i::Code> wrapper =
